src/tools/gravity-parseList.c

```diff
@@ -330,162 +330,160 @@ int gravity_parseList(const char *infile, const char *outfile, const char *adlis
+					fclose(fpin);
+					sqlite3_close(db);
+					return EXIT_FAILURE;
+				}
+				sqlite3_reset(stmt);
+				abp_domains++;
+			}
+			else
+			{
+				// No match - This is an invalid domain or a false positive
 
-						// Compare against the current domain
-						if(memcmp(invalid_domains_list[i], line, min(read, invalid_domains_list_lengths[i])) == 0)
+				// Ignore false positives - they don't count as invalid domains
+				if(!is_false_positive(token))
+				{
+					if(checkOnly)
+					{
+						// Increment counter
+						invalid_domains++;
+						printf("%s  %s Invalid domain on line %zu: ", over, cross, lineno);
+						print_escaped(token, token_len);
+						puts("");
+						goto next_domain;
+					}
+					// Add the domain to invalid_domains_list only
+					// if the list contains < MAX_INVALID_DOMAINS
+					if(invalid_domains_list_len < MAX_INVALID_DOMAINS)
+					{
+						// Check if we have this domain already
+						bool found = false;
+						for(unsigned int i = 0; i < invalid_domains_list_len; i++)
 						{
-							found = true;
-							break;
+							// Do not compare against unset entries
+							if(invalid_domains_list[i] == NULL || invalid_domains_list_lengths[i] == -1)
+								break;
+
+							// Compare against the current domain
+							if(memcmp(invalid_domains_list[i], token, min((ssize_t)token_len, invalid_domains_list_lengths[i])) == 0)
+							{
+								found = true;
+								break;
+							}
 						}
-					}
 
-					// If not found, add it to the list
-					if(!found)
-					{
-						invalid_domains_list[invalid_domains_list_len] = calloc(read + 1, sizeof(char));
-						if(invalid_domains_list[invalid_domains_list_len] == NULL)
+						// If not found, add it to the list
+						if(!found)
 						{
-							printf("%s  %s Unable to allocate memory for invalid domains list\n", over, cross);
-							fclose(fpin);
-							sqlite3_close(db);
-							return EXIT_FAILURE;
+							invalid_domains_list[invalid_domains_list_len] = calloc(token_len + 1, sizeof(char));
+							if(invalid_domains_list[invalid_domains_list_len] == NULL)
+							{
+								printf("%s  %s Unable to allocate memory for invalid domains list\n", over, cross);
+								fclose(fpin);
+								sqlite3_close(db);
+								return EXIT_FAILURE;
+							}
+							memcpy(invalid_domains_list[invalid_domains_list_len], line, token_len);
+							invalid_domains_list[invalid_domains_list_len][token_len] = '\0';
+							invalid_domains_list_lengths[invalid_domains_list_len] = token_len;
+							invalid_domains_list_len++;
 						}
-						memcpy(invalid_domains_list[invalid_domains_list_len], line, read);
-						invalid_domains_list[invalid_domains_list_len][read] = '\0';
-						invalid_domains_list_lengths[invalid_domains_list_len] = read;
-						invalid_domains_list_len++;
-					}
 
+					}
+					invalid_domains++;
 				}
-				invalid_domains++;
 			}
+next_domain:
+			token = strtok(NULL, " \t");
```

Comment:
Why do you need to split again here? We're still in the loop, we just need to go to the next iteration.

Reply:
We split once *before* the loop. We then split again at the end of the loop to get the next element. If this turns out to not exist (have handled everything), `strtok` returns `NULL` which will end the `while(token != NULL)` loop.